TNL-4325: Linter Enhancements for JavaScript files

robrap opened the pull request:
## [TNL-4325](https://openedx.atlassian.net/browse/TNL-4325)

Add JavaScript and CoffeeScript rules

- Check for concatenating strings with HTML.
- Checks various jQuery functions
- Checks CoffeeScript
- Checks underscore templates embedded in JavaScript and CoffeeScript
- Checks interpolate() calls
- Checks escape() calls

Sample output from the latest linter with the new rule:
```
cms/static/js/factories/export.js: 36:20: javascript-concat-html:             msg += '</p><p>' + gettext('The raw error message is:') + '</p>' + errMsg;
cms/static/js/factories/import.js: 33:31: javascript-jquery-html:                 .find('.copy').html(gettext("Choose new file")).end()
cms/static/js/models/settings/course_details.js: 62:57: javascript-interpolate:                 errors.entrance_exam_minimum_score_pct = interpolate(gettext("Please enter an integer between %(min)s and %(max)s."), range, true);
cms/static/js/views/course_info_update.js: 81:58: javascript-jquery-append:                     $(ele).find('.date-display').parent().append(this.errorTemplate({message : error[field]}));
cms/static/js/views/video/translations_editor.js: 121:24: javascript-jquery-prepend:                     })).prepend(dropdown.clone().val(key))[0];
cms/static/js/views/pages/container.js: 212:73: javascript-jquery-insert-into-target:                     placeholderElement = this.createPlaceholderElement().appendTo(listPanel),
cms/static/js/views/pages/container.js: 322:43: javascript-jquery-insertion:                         this.$el = $(html).replaceAll(element);
common/lib/xmodule/xmodule/js/src/problem/edit.coffee: 543:30: javascript-concat-html:           var selectString = '<pre><code>\n' + p1 + '</code></pre>';
cms/static/js/views/edit_chapter.js: 63:38: javascript-escape:                         {name: course.escape('name')}),
```

### Testing
- [x] Data is properly encoded in HTML templates to avoid XSS
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @efischer19 
- [x] Code review: @nedbat 
- [x] Code review: @andy-armstrong 

### Post-review
- [x] Squash commits

robrap commented on scripts/safe_template_linter.py:
```diff
@@ -1254,6 +1272,202 @@ def _find_unescaped_expressions(self, underscore_template):
+            "$". For example, "testEl" or "$test".
+        - the argument can be a single string literal with no HTML tags
+        - the argument can start with "[" or "{". If so, we assume this is an
+            array or object append() call.
```
Hi @andy-armstrong.  I'm trying to not have the linter complain in places where it doesn't need to complain. 

For the the rule above, I now realize that JavaScript objects and arrays don't have an append() function.  However, I'm not quite sure what is going on in some of these lines:
https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/progress_graph.js#L50-51
https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/progress_graph.js#L83-L87
Are these still all just JQuery append?  Some of these look like plain old arrays, so I was confused.  Should I just drop this exception altogether?

robrap commented on scripts/safe_template_linter.py:
```diff
@@ -1254,6 +1272,202 @@ def _find_unescaped_expressions(self, underscore_template):
+            array or object append() call.
+        - the argument can be a call to $() with the first argument a string
+            literal with a single HTML tag.  For example, ".append($('<br/>'))"
+            or ".append($('<br/>'))".
```
@andy-armstrong I decided to skip these in the linter because it appeared a bunch, and seems safe.  Does that seem ok to you?  Reviewing the tests might make these more clear.

robrap commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -730,3 +737,53 @@ def test_check_underscore_no_escape_allowed(self, data):
+        {'template': 'test.append( $( "<div/>" ) )', 'rule': None},
+        {'template': 'test.append( [] )', 'rule': None},
+        {'template': 'test.append( {x: 1, y: 2 + 3} )', 'rule': None},
+        {'template': 'HtmlUtils.append(anything)', 'rule': None},
```
@andy-armstrong The tests above all show what is deemed safe.  As mentioned above, it might make sense to drop the rule for assuming arrays and objects are safe, but I'm not sure how these get used.  Are these non-JQuery appends (see examples above)?

robrap replied:
I've since added `test.append(HtmlUtils.forJQuery(anything))` as being acceptable as well.

robrap commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -730,3 +737,53 @@ def test_check_underscore_no_escape_allowed(self, data):
+        {'template': 'test.append( test.render().$el )', 'rule': None},
+        {'template': 'test.append( testEl )', 'rule': None},
+        {'template': 'test.append( testEl )', 'rule': None},
+        {'template': 'test.append( $test )', 'rule': None},
```
I currently have a rule to accept single variables with names that match the convention `anythingEl` or `$anything`. Does this make sense to you?

robrap commented:
@andy-armstrong This is linting just one of the JQuery calls so far (i.e. append).  I've added notes so we can discuss the rule and possible exceptions.

Thank you.

robrap commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -721,12 +726,195 @@ def test_check_underscore_file_disables_one_violation(self):
+
+
+@ddt
+class TestJavaScriptLinter(TestLinter):
```
@andy-armstrong @efischer19 @nedbat: FYI: I'd start by reviewing the tests to understand what is being linted for.

andy-armstrong replied:
I'd like to see some Unicode text included in these tests to verify that it doesn't cause unexpected behavior. I would make any strings include some random Unicode characters, e.g. 檌檒濦

robrap commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -721,12 +733,197 @@ def test_check_underscore_file_disables_one_violation(self):
+        self._validate_data_rule(data, results)
+
+    @data(
+        {'template': 'test.prepend( test.render().el )', 'rule': None},
```
@andy-armstrong: I added linting of CoffeeScript too.  Is there anything like `@el` that should be exempt?  Maybe you could add a comment with some tests that you think should pass in CoffeeScript.  This may apply to some other tests as well.  I don't know CoffeeScript.  Thanks!

andy-armstrong replied:
I scanned through CoffeeScript calls with append, prepend etc, and didn't come across any variants that are worth adding. The syntax is just the same as for JavaScript.

andy-armstrong commented on scripts/safe_template_linter.py:
```diff
@@ -56,6 +56,79 @@ def _load_file(self, file_full_path):
+    """
+    strings = [] if strings is None else strings
+    unparseable_result = {'close_char_index': -1, 'strings': []}
+    close_char_index = template.find(close_char, start_index)
```
How would this handle a closing character inside a string before the real closing character? Would it match it incorrectly? e.g. when matching parens, what would happen with:

    this.$el.text('Exit Music (For a Film)');

robrap replied:
It deals intelligently with inner function calls and strings.

andy-armstrong commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -721,12 +733,197 @@ def test_check_underscore_file_disables_one_violation(self):
+        {'template': 'test.append( "text" )', 'rule': None},
+        {'template': 'test.append( $( "<div>" ) )', 'rule': None},
+        {'template': 'test.append( $( "<div/>" ) )', 'rule': None},
+        {'template': 'test.append(HtmlUtils.forJQuery(someHtml))', 'rule': None},
```
Reminder to remove these now that we've decided against ``forJQuery``.

robrap replied:
@andy-armstrong: I adjusted the rules.  See the latest commit.

andy-armstrong commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -721,12 +733,197 @@ def test_check_underscore_file_disables_one_violation(self):
+        self._validate_data_rule(data, results)
+
+    @data(
+        {'template': 'test.append( test.render().el )', 'rule': None},
```
Do you need tests that don't add the extra spaces too? Most JavaScript styles would not include such spaces.

robrap replied:
I fixed the tests to have some with extra spaces, but defaulted to typical syntax.

andy-armstrong commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -721,12 +733,197 @@ def test_check_underscore_file_disables_one_violation(self):
+        {'template': 'test.append( test.render().$el )', 'rule': None},
+        {'template': 'test.append( testEl )', 'rule': None},
+        {'template': 'test.append( $test )', 'rule': None},
+        {'template': 'test.append( "text" )', 'rule': None},
```
Shouldn't this fail because the string isn't HTML escaped?

robrap replied:
I added some explanation.  This is an odd case, but it is safe and gets rid of some false positives on other objects that have an append() method and take a string.

andy-armstrong commented on scripts/tests/test_safe_template_linter.py:
```diff
@@ -721,12 +733,197 @@ def test_check_underscore_file_disables_one_violation(self):
+        self._validate_data_rule(data, results)
+
+    @data(
+        {'template': 'test.unwrap(HtmlUtils.forJQuery(someHtml))', 'rule': None},
```
Remove all these ``forJQuery`` tests

andy-armstrong replied:
I suppose I mean replace them with the ``toString`` version.

robrap commented:
@andy-armstrong: I think I addressed all your concerns.  

My main question/hesitation is around when we use naming conventions for acceptable arguments.  We still think that considering `test.append(testEl)' and `test.append($test)` to be safe is a reasonable way to go, right?

andy-armstrong commented:
@robrap It's a good question. There will be a lot of failures if we don't do this, but I agree with your concern that it could allow bad code to slip through. Do you have a way to disable linting on a line? If so, then maybe we should try without these conventions and see how bad it is.

robrap commented:
@andy-armstrong: I had missed your unicode concern, so I just added 2 tests.

robrap commented:
@andy-armstrong: I only found 12 errors that would be found if I dropped the exceptions for `test.append(testEl)` and `test.append($test)`.  

We do have a disable pragma.  However, I also don't like that there isn't a way to code this appropriately to start.  Without this naming convention exception, I guess the alternative is:
```
testHtmlString = testEl.html()
test.append(HtmlUtils.HTML(testHtmlString).toString())
```

I think I prefer relying on the naming convention.  Do you think both these conventions are equally risky/unrisky?  I'm leaning towards leaving it as currently implemented, but how do you feel now?

efischer19 commented:
👍 This all seems reasonable to me. Good test coverage!

andy-armstrong commented:
@robrap I definitely don't like the idea of converting an element into a string in order to append it. In Backbone, we often pass one view as the child of another, and so it needs to use the view's own element. Given this, I agree with leaving the current conventions in place.

andy-armstrong commented:
👍  from me. You should rebase to pick up the changes that fix the JavaScript test failures.

robrap commented:
Thanks @andy-armstrong and @efischer19.  I'll do that when I squash, which will be after I merge in my other two PRs as well.  Thanks!

robrap commented:
jenkins run lettuce
